test/test_pumactl.rb

```diff
@@ -132,24 +135,53 @@ def test_control_url_and_status
 
     wait_booted
 
-    s = TCPSocket.new host, 9292
-    s << "GET / HTTP/1.0\r\n\r\n"
-    body = s.read
-    assert_match "200 OK", body
-    assert_match "embedded app", body
+    assert_app_running host
+    assert_command_cli_output opts + ["status"], "Puma is started"
+    assert_command_cli_output opts + ["stop"], "Command stop sent success"
 
-    status_cmd = Puma::ControlCLI.new(opts + ["status"])
-    out, _ = capture_subprocess_io do
-      status_cmd.run
+    assert_kind_of Thread, t.join, "server didn't stop"
+  end
+
+  def test_control_ssl
+    host = "127.0.0.1"
+    port = find_open_port
+    url = "ssl://#{host}:#{port}?#{ssl_query}"
+
+    opts = [
+      "--control-url", url,
+      "--control-token", "ctrl",
+      "--config-file", "test/config/app.rb",
+    ]
+
+    control_cli = Puma::ControlCLI.new (opts + ["start"]), @ready, @ready
+    t = Thread.new do
+      control_cli.run
     end
-    assert_match "Puma is started\n", out
 
-    shutdown_cmd = Puma::ControlCLI.new(opts + ["halt"])
+    wait_booted
+
+    assert_app_running host
+    assert_command_cli_output opts + ["status"], "Puma is started"
+    assert_command_cli_output opts + ["stop"], "Command stop sent success"
+
+    assert_kind_of Thread, t.join, "server didn't stop"
+  end
+
+  private
+
+  def assert_command_cli_output(options, expected_out)
+    cmd = Puma::ControlCLI.new(options)
     out, _ = capture_subprocess_io do
-      shutdown_cmd.run
+      cmd.run
     end
-    assert_match "Command halt sent success\n", out
+    assert_match expected_out, out
+  end
 
-    assert_kind_of Thread, t.join, "server didn't stop"
+  def assert_app_running(host)
```

Comment:
I'm not sure this is a useful abstraction, because you're not actually asserting that "an app" is "running,", you're asserting that a puma server running the "embedded app" config file returns a 200 when connected to.

Reply:
Doesn't wait_booted perform a check something like this, as well?

Reply:
Good point. This code didn't seem important to what either test was actually testing, so I tried to hide it away in a helper method. Digging further, I see now that `wait_booted` on was waiting for the embedded app to boot, but not for the control app. Connecting to the "embedded app" took long enough that the control app was booted by the time we needed it. `sleep 0.1` would also have worked.

I am going to update `wait_booted` (used only in these two tests) to wait for the control app instead, and then get rid of the "embedded app" assertion for the new spec. I will the extra assertions it in the existing spec in case they are testing something that isn't covered elsewhere.